### ISSUE TITLE
Please use timestamps in backup file names instead of tilde to reduce number of syscalls

### ISSUE THREAD

**Issue by mdirik** [giuspen/cherrytree#1470]
**Background**

I set autosave feature to save files frequently, and I have a high number (100) set for the "number of backups to keep" option so that the backup files can go back few days.

**Problem**

I noticed the file scheme for backup files is appending `~` to the main file name, and every older backup file also needs to be renamed so that the most recent backup can have the name `<file>.ctb~`.

This causes high amount of syscall activity on the system because all the older file names must be checked for existence and renamed.

This is the trace of syscalls that are called for renaming 1 backup file:

```
stat64("/home/user/cherrytree-tilda-backup/test.ctb~~~~~~~~", {st_mode=S_IFREG|0644, st_size=45056, ...}) = 0
lstat64("/home/user/cherrytree-tilda-backup/test.ctb~~~~~~~~", {st_mode=S_IFREG|0644, st_size=45056, ...}) = 0
lstat64("/home/user/cherrytree-tilda-backup/test.ctb~~~~~~~~~", 0xffc1074c) = -1 ENOENT (No such file or directory)
rename("/home/user/cherrytree-tilda-backup/test.ctb~~~~~~~~", "/home/user/cherrytree-tilda-backup/test.ctb~~~~~~~~~") = 0
lstat64("/home/user/cherrytree-tilda-backup", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat64("/home/u1", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat64("/home", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat64("/", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
eventfd2(0, EFD_CLOEXEC|EFD_NONBLOCK)   = 13
close(13)                               = 0
```

Same is repeated for every single old backup file (up to 100 if the relevant option is set to that value).

Considering filesystem relevant syscalls are expensive, this becomes a problem when autosave kicks in. The slow down can be easily spotted when the # of backups is high and the backups are stored on a slow media such as hard drives.

**Describe the solution you'd like**

The easiest solution is using timestamped file names for backup files. A possible example to that would be `document.ctb.202101040114`.

This would only require creating a new file, and alleviate the need for checking for and renaming all the older files sequentially.

This creates a new issue, for figuring out the names of older backup files to keep them under the # of backup files limit, but my guess is it would still be less expensive operation than calling `rename` on succession.

Thank you very much for this great piece of software and have a nice day!

**Comment by datavectors** [comment]
I am just a user of CherryTree on Ubuntu desktop.
100 backups is quite a lot. If that degree of backup is needed (no rolling backups) then one thought I have is to run some cron script which:
analyses CT backup files
rewrites each one as *.ctd (xml) file
pushes the backup (xml) as document collection to a NoSQL database such as eXist-db.
deletes the desktop backup files (ctb and ctd)

This would give a database for audit purposes.
A refinement might be to just backup _differences_ (rdiff) in files to give an audit trail.
Another refinement might be to just hold a few (4?) conventional backups and overflows to be pushed to database. A two stage process.

Just some initial ideas I would pursue if I needed that granularity of backup. Certainly not 100 tildes in file names!

**Comment by steveno** [comment]
I agree with @datavectors that @mdirik's use case is extreme, but I think that @mdirik's method is much better than the tildas. The potential pitfalls there, I feel like, should be manageable, and the upsides to it would be worth it.

**Comment by giuspen** [comment]
It makes sense, I'll do something
